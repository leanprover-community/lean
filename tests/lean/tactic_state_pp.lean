universe variables u

inductive Vec (α : Type u) : nat → Type u
| nil  : Vec 0
| cons : ∀ {n}, α → Vec n → Vec (nat.succ n)

constant f {α : Type u} {n : nat} : Vec α n → nat
axiom fax1 (α : Type u) : f (Vec.nil : Vec α 0) = 0
axiom fax2 {α : Type u} {n : nat} (v : Vec α (nat.succ n)) : f v = 1

open tactic
meta def pp_state_core : tactic format :=
do t     ← target,
   t_fmt ← pp t,
   return $ to_fmt "Goal: " ++ t_fmt

meta def pp_state (s : tactic_state) : format :=
match pp_state_core s with
| result.success r _     := r
| result.exception _ _ _ := "failed to pretty print"
end

meta instance i2 : has_to_format tactic_state :=
⟨λ s, to_fmt "My custom goal visualizer" ++ format.line ++ pp_state s⟩

example {α : Type u} {n : nat} (v : Vec α n) : f v ≠ 2 :=
begin
  destruct v,
  intros, intro, have h := fax1 α, cc,
  -- intros n1 h t, intros, intro, have h := fax2 (Vec.cons h t), cc
end

open nat
example : ∀ n, 0 < n → succ (pred n) = n :=
begin
  intro n,
  destruct n,
   dsimp, intros, have h := lt_irrefl 0, cc,
end
