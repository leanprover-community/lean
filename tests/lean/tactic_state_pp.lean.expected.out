tactic_state_pp.lean:31:0: error: tactic failed, there are unsolved goals
state:
My custom goal visualizer
Goal: ∀ {n_1 : ℕ} (a : α) (a_1 : Vec α n_1),
  (λ (a : ℕ) (w : Vec α a), n = a → v == w → f v ≠ 2) n_1.succ (Vec.cons a a_1)
tactic_state_pp.lean:39:0: error: tactic failed, there are unsolved goals
state:
My custom goal visualizer
Goal: ∀ (n_1 : ℕ), (λ (w : ℕ), n = w → 0 < n → n.pred.succ = n) n_1.succ
